File: 9_PalindromeNumber.js
/**
 * @param {number} x
 * @return {boolean}
 */
//解法1分別從左邊即右邊遍歷組成數字  相減=0即為迴文
const isPalindrome = function (x) {
    //負數一定不為迴文
    if (x < 0) return false;
    let result = false;

    let str = x.toString();
    //left為左向右組成的數字  反之
    let left = '', right = '';
    //由左向右遍歷
    for (let i = 0; i < str.length; i++) {
        left += str[i];
    }
    //由右向左遍歷
    for (let i = str.length - 1; i >= 0; i--) {
        right += str[i];
    }
    result = (Number(left) - Number(right)) === 0 ? true : false;
    return result;
};

//解法2 遞迴從左右端點遍歷比對 每次遞迴僅比對頭尾端點
const isPalindrome2 = function (x) {
    const str = x.toString();
    //長度 === 1  代表傳進字串為迴文
    if (str.length === 1) return true;
    //起始索引
    const start = str[0];
    //最後索引
    const end = str[str.length - 1];

    if (start !== end) return false;

    //若是經過前面判斷尚未回傳false  此時若字串長度為2  此字串即為迴文
    if (str.length === 2) return true;

    //比對完成切割字串
    let subStr = str.substring(1, str.length - 1);
    return isPalindrome2(subStr);
}

File: 09.Palindrome Number/9.Palindrome Number.py
# 解法1 分別從左邊即右邊遍歷組成數字  相減=0即為迴文
def isPalindrome(x: int) -> bool:
    # 負數一定不為迴文
    if x < 0:
        return False
    # 轉換成字串
    string = str(x)
    # left right分別為string從左到右及從右到左的組成的結果
    left, right = '', ''
    # 左至右
    for i in range(len(string)):
        left += string[i]
    # 右至左
    for j in range(len(string) - 1, -1, -1):
        right += string[j]

    return (int(left) - int(right)) == 0

# 解法2 follow up:不能轉成字串
# 可以利用取整和取餘來獲得想要的數位，
# 比如 1221 這個數位，如果 計算 1221 / 1000， 則可得首位1，
# 如果 1221 % 10， 則可得到末尾1，進行比較，然後把中間的 22 取出繼續比較
def isPalindrome1(x: int) -> bool:
    if x < 0:
        return False
    # 用來取整數的值
    div = 1
    while x / div >= 10:
        div *= 10
    while x > 0:
        left = x // div
        right = x % 10
        if left != right:
            return False
        # x去掉頭尾的值  且一次去掉兩個數故div/100
        x = (x % div) // 10
        div /= 100

    return True

isPalindrome1(121)